cmd/bosun/web/static/js/config.ts

```diff
@@ -1,55 +1,305 @@
 	var search = $location.search();
-	var current = search.config_text;
-	var line_re = /test:(\d+)/;
-	try {
-		current = atob(current);
+	$scope.fromDate = search.fromDate || '';
+	$scope.fromTime = search.fromTime || '';
+	$scope.toDate = search.toDate || '';
+	$scope.toTime = search.toTime || '';
+	$scope.intervals = +search.intervals || 5;
+	$scope.duration = +search.duration || null;
+	$scope.config_text = 'Loading config...';
+	$scope.selected_alert = search.alert || '';
+	$scope.email = search.email || '';
+	$scope.template_group = search.template_group || '';
+	$scope.items = parseItems();
+	$scope.tab = search.tab || 'results';
+	
+	var textElem = $("#configText");
+	$timeout(()=>{ 
+			textElem.linedtextarea();
+	})
+	
+	
+	var expr = search.expr
+	
+	function buildAlertFromExpr(){
+		if (!expr) return
+		var newAlertName = "test";
+		var idx = 1;
+		//find a unique alert name
+		while($scope.items["alert"].indexOf(newAlertName) != -1 || $scope.items["template"].indexOf(newAlertName) != -1){
+			newAlertName = "test" + idx;
+			idx++;
+		}
+		var text = '\n\ntemplate '+newAlertName+' {\n' +
+			'	subject = {{.Last.Status}}: {{.Alert.Name}} on {{.Group.host}}\n' +
+			'	body = `<p>Name: {{.Alert.Name}}\n' +
+			'	<p>Tags:\n' +
+			'	<table>\n' +
+			'		{{range $k, $v := .Group}}\n' +
+			'			<tr><td>{{$k}}</td><td>{{$v}}</td></tr>\n' +
+			'		{{end}}\n' +
+			'	</table>`\n' +
+			'}\n\n';
+		text += 'alert '+newAlertName+' {\n' +
+			'	template = '+newAlertName+'\n' +
+			'	crit = ' + atob(expr) + '\n' +
+			'}\n';
+		$scope.config_text += text;
+		$scope.items = parseItems();
+		$timeout(()=>{ 
+				//can't scroll editor until after control is updated. Defer it.
+				$scope.scrollTo("alert",newAlertName)
+		})
 	}
-	catch (e) {
-		current = '';
+		
+	
+	function parseItems() : { [type: string]: string[]; }{
+		var configText = $scope.config_text;
+		var re = /^\s*(alert|template|notification|lookup|macro)\s+([\w\-\.\$]+)\s*\{/gm; 
+		var match;
+		var items : { [type: string]: string[]; } = {};
+		items["alert"] = [];
+		items["template"] = [];
+		items["lookup"] = [];
+		items["notification"] = [];
+		items["macro"] = [];
+		while (match = re.exec(configText)) {
+        		var type = match[1];
+			var name = match[2];
+			var list = items[type];
+			if (!list){
+				list = [];
+				items[type] = list;
+			}
+			list.push(name)
+		}
+		return items
 	}
-	if (!current) {
-		var def = '';
-		$http.get('/api/config')
+	
+	$http.get('/api/config')
+		.success((data) => {
+			$scope.config_text = data;
+			$scope.items = parseItems();
+			buildAlertFromExpr()
+			if(!$scope.selected_alert && $scope.items["alert"].length){
+				$scope.selected_alert = $scope.items["alert"][0];
+			}
+			$timeout(()=>{ 
+				//can't scroll editor until after control is updated. Defer it.
```

Comment:
oh i think this is one reason we had the tsLine directive: directives are able to trigger at the correct time so this is possible. or maybe i'm remembering it wrong, but that sounds right. i forget.